File: Stacks/StackLL.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

// Linked List
typedef struct Node{
    int data;
    struct Node* next;
} Node;

//Stack
typedef struct Stack {
    Node* top;
} Stack;

void create(Stack* st){
    st->top = NULL;
}

bool isEmpty(Stack st){
    return st.top == NULL;
}

bool push(Stack* st, int val){
    Node* t = malloc(sizeof(Node));
    if(!t){
        printf("Stack overflow!");
        return false;
    }
    
    t->data = val;
    t->next = st->top;
    st->top = t;
    
    return true;
}

int pop(Stack* st){
    int x = -1;
    if(isEmpty(*st)){
        printf("Stack is empty!");
        return x;
    }
    
    Node* t = st->top;
    st->top = st->top->next;
    
    x = t->data;
    
    free(t);
    t = NULL;
    
    return x;
}

int peek(Stack st, int pos){
    Node* t = st.top;
    for(int i = 0; t && i < pos; i++){
        t = t->next;
    }
    
    if(t) return t->data;
    
    return -1;
}

int stack_top(Stack st){
    if(!isEmpty(st)) return st.top->data;
    
    return -1;
}

void display(Stack st){
    Node* t = st.top;
    while(t != NULL){
        printf("\n%d", t->data); 
        t = t.next;   
    }
}

int main(void){
    
    return 0;
}

File: Stacks/StackLL.java
public class StackLL <T>{
    
    private class Node {
        private T data;
        private Node next;
        
        public Node(T data, Node next){
            this.data = data;
            this.next = next;
        }
        
        
    }
    
    private Node top;
    
    public StackLL(){
        this.top = null;
    }
    
    public boolean isEmpty(){
        return this.top == null;
    }
    
    public boolean push(T val){
        this.top = new Node(val, top);
        return true;
    }
    
    public T pop(){
        if(this.isEmpty()){
            System.out.println("Stack is empty! Nothing to Pop");
            return null;
        }
        
        T popped = this.top.data;
        
        top = top.next;
        
        return popped;
    }
    
    public T peek(int position){
        
        if(this.isEmpty()){
            System.out.println("Stack is empty! Nothing to Peek");
            return null;
        }
        
        if(position < 0)
            throw new IllegalArgumentException("Invalid index!");
        
        Node t = top;
        for(int i = 0; t != null && i < position; i++){
            t = t.next;
        }
        
        if(t == null)
            throw new IllegalArgumentException("Position out of bounds!");
        
        return t.data;
    }
    
    public T stackTop(){
        if(isEmpty()){
            System.out.println("Stack is empty!");
            return null;
        }
        
        return this.top.data;
    }
    
    public int size(){
        int size = 0;
        
        Node t = top;
        while(t != null){
            size++;
            t = t.next;
        }
        
        return size;
    }
    
}